ramips: add support for TOZED ZLT S12 PRO

arinc9 opened the pull request:
Hello, this PR adds support for the TOZED LT70-C cellular modem and TOZED ZLT S12 PRO router. I am also adding the vendor name to Lenovo Newifi D1 and D-Team Newifi D2 devices.

arinc9 committed:
ramips: add support for TOZED ZLT S12 PRO

The TOZED ZLT S12 PRO is an AC1200 router featuring 4 Ethernet ports with a
TOZED TL70-C cellular modem which supports the NCM mode.

The stock firmware does SIM locking on the modem by stopping dialing when a
different PLMN is detected. This is not the case on OpenWrt.

Specifications:
- CPU: MediaTek MT7621AT
- RAM: 256MB DDR3
- NOR Flash: MX25L12833FM2I 16MB SPI Flash
- Wi-Fi 2.4Ghz: MT7603E
- Wi-Fi 5Ghz: MT7612E
- Switch: MT7530 4x 1Gbit Ports
- WWAN: Unisoc SL8563 based TOZED TL70-C LTE CAT6 cellular modem
- USB: 1x optional USB2.0 external port
- Switches/Buttons: WPS, Reset, Power Switch
- LEDs: Power, Wi-Fi, Data, Signal 1-5, Phone

Installation and TFTP Recovery:
- Connect to serial console.
- Boot initramfs image by choosing option 1 when U-Boot prompts.
- Install sysupgrade image via OpenWrt.

Serial Pins:
Located at the bottom right when looking from the front, right under the
Reset/WPS buttons. The pinout from the left is:
- RX
- GND
- TX
Baudrate is 115200.

When connecting from a powered off state, disconnect RX as it blocks the
boot process.

Link: http://www.sztozed.com/en/contents/58/84.html
Co-developed-by: Andre Cruz <me@1conan.com>
Signed-off-by: Andre Cruz <me@1conan.com>
Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

arinc9 committed:
comgt-ncm: add support for UNISOC SL8563 based TOZED TL70-C

TOZED TL70-C is an LTE CAT6 cellular modem based on UNISOC SL8563. UNISOC
was formerly called Spreadtrum hence the manufacturer name detected on the
modem is spreadtrum.

The connect and disconnect commands bring up and down the usb0 interface.
They are Base64 encoded as that's what the AT command accepts. The modem
can do up to 4 APNs by bringing the USB interfaces, usb0 to usb3, up.

Setting the USB interfaces up:

connmanctl ndisdial AT^NDISDUN="usb0",1,1
connmanctl ndisdial AT^NDISDUN="usb1",1,2
connmanctl ndisdial AT^NDISDUN="usb2",1,3
connmanctl ndisdial AT^NDISDUN="usb3",1,4

Setting the USB interfaces down:

connmanctl ndisdial AT^NDISDUN="usb0",0,1
connmanctl ndisdial AT^NDISDUN="usb1",0,2
connmanctl ndisdial AT^NDISDUN="usb2",0,3
connmanctl ndisdial AT^NDISDUN="usb3",0,4

Co-developed-by: Andre Cruz <me@1conan.com>
Signed-off-by: Andre Cruz <me@1conan.com>
Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

arinc9 committed:
ramips: rename to D-Team Newifi D2 and Lenovo Newifi D1

Rename Newifi D2 to D-Team Newifi D2, and Newifi D1 to Lenovo Newifi D1.
Let's not make naming exceptions because of marketing whims.

Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

arinc9 commented:
@hauke looks good?

hauke commented on target/linux/ramips/dts/mt7621_tozed_zlt-s12-pro.dts:
```diff
@@ -0,0 +1,245 @@
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
```
Is it possible to access this u-boot environment from Linux?

arinc9 replied:
This is doable with using the uboot-envtools package. The variables won't appear properly on this specific device, so I'm not going to include that.

I also think the `NVMEM_U_BOOT_ENV` driver is the way to go forward which lets you define the uboot-env partition on the devicetree instead of the OpenWrt-specific `package/boot/uboot-envtools/files/`. The fw_setenv and fw_printenv commands from the uboot-envtools package should start using the driver in the future. The driver hasn't got writing ability (@reg_write) as of now so that needs to be addressed first. @rmilecki let me know your thoughts.

arinc9 replied:
I see it is already implemented. https://github.com/u-boot/u-boot/commit/8142c4554ffaa927529f24427a35f7ee2861793a

We should make the uboot-envtools package on OpenWrt follow suit.

hauke replied:
The patch should be included in the u-boot envtools version we are using in master. I have never tried how it works. Maybe we have to adapt some scripts to make it fully work in OpenWrt.

hauke commented:
When you add such an entry:
````
ucidef_set_interface "wan" device "/dev/cdc-wdm0" protocol "qmi"
````
The following entry will be created at first bootup:
````
config interface 'wan'
	option device '/dev/cdc-wdm0'
	option proto 'qmi'
	option pdptype 'ipv4'
````
LuCI looks like this:
![image](https://github.com/openwrt/openwrt/assets/78494/1fc9dc65-6a7c-4043-a0a7-f2141941a8d2)

It looks like ncm is not yet supported, see https://github.com/openwrt/openwrt/blob/master/package/base-files/files/bin/config_generate#L211

arinc9 commented:
Looks like @1Conan's comment above could fairly easily be implemented. What do you think Hauke?

arinc9 committed:
base-files: add protocol ncm support for ucidef_set_interface()

Make it possible to set up a default network (interface) for devices with
cellular modems using NCM.

Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

arinc9 commented:
@1Conan could you test the current PR? I didn't set apn to "internet" as it could be something else by other carriers? Let me know. pdptype is ipv4 instead of "IP", like you specified. Make sure it works with ipv4.

arinc9 commented:
I changed the co-authored-by tag with co-developed-by as the prior seems to be a GitHub specific tag whilst the latter is the only documented one on the submitting patches page on the linux kernel documentation.
